src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -48,5 +50,129 @@ public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private Dictionary<Tuple<string, string>, EventBase> EventsInProgress = new Dictionary<Tuple<string, string>, EventBase>();
```

Comment:
You mean use a thread-safe [`ConcurrentDictionary`](https://msdn.microsoft.com/en-us/library/dd287191(v=vs.110).aspx), like what [Android Telemetry already did](https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/90/files#diff-06c1d15c8ab55d113eb9d700468298ffR57)? OK we can do that.